lib/charms/vault_k8s/v0/vault_autounseal.py

```diff
@@ -0,0 +1,371 @@
+    """Events raised by the vault-autounseal relation on the provider side."""
+
+    vault_autounseal_initialize = ops.EventSource(VaultAutounsealInitialize)
+    vault_autounseal_destroy = ops.EventSource(VaultAutounsealDestroy)
+
+
+class VaultAutounsealRequireEvents(ops.ObjectEvents):
+    """Events raised by the vault-autounseal relation on the requirer side."""
+
+    vault_autounseal_details_ready = ops.EventSource(VaultAutounsealDetailsReadyEvent)
+
+
+class VaultAutounsealProvides(ops.Object):
+    """Manages the vault-autounseal relation from the provider side."""
+
+    on: VaultAutounsealProvidesEvents = VaultAutounsealProvidesEvents()  # type: ignore
+
+    def __init__(self, charm: ops.CharmBase, relation_name: str):
+        super().__init__(charm, relation_name)
+        self.charm = charm
+        self.relation_name = relation_name
+
+        self.framework.observe(
+            self.charm.on[relation_name].relation_created, self._on_relation_created
+        )
+        self.framework.observe(
+            self.charm.on[relation_name].relation_broken, self._on_relation_broken
+        )
+
+    def _on_relation_created(self, event: ops.RelationCreatedEvent) -> None:
+        self.on.vault_autounseal_initialize.emit(relation=event.relation)
+
+    def _on_relation_broken(self, event: ops.RelationBrokenEvent) -> None:
+        self.on.vault_autounseal_destroy.emit(relation=event.relation)
+
+    def set_vault_url(self, relation: ops.Relation, address: str) -> None:
+        """Set the vault url address in the relation databag.
+
+        Args:
+            relation: The relation for which to set the vault url.
+            address: The address of the vault server.
+
+        """
+        if not self.charm.unit.is_leader():
+            logger.warning(
+                "Attempting to set the vault url without being the leader. Ignoring the request."
+            )
+            return
+        relation.data[self.charm.app].update({"address": address})
+
+    def set_credentials_secret_id(self, relation: ops.Relation, secret_id: str) -> None:
+        """Set the credentials secret id in the relation databag.
+
+        Args:
+            relation: The relation for which to set the credentials secret id.
+            secret_id: The secret id of the Juju secret which stores the credentials for authenticating with the Vault server.
+        """
+        if relation.app is None:
+            raise ValueError("The `relation.app` is not set")
+        relation.data[self.charm.app].update({"credentials_secret_id": secret_id})
+
+    def set_ca_certificate(self, relation: ops.Relation, ca_certificate: str) -> None:
+        """Set the ca_certificate on the relation.
+
+        Args:
+            relation: The relation for which to set the ca_certificate.
+            ca_certificate: The CA certificate to use when validating the Vault server's certificate.
+        """
+        if not self.charm.unit.is_leader():
```

Comment:
Why do we have different validations here vs. in `set_credentials_secret_id` for example?

Reply:
Once again, mostly just modelled off how `vault_kv` was doing things.

I admittedly don't have a great intuition on what the edge cases are for most of these, so I tried to conform.

Reply:
fixed in https://github.com/canonical/vault-k8s-operator/pull/346/commits/0651467cd3a75065ba0f04391194ceef96844526